File: src/ImputeNoRef.cpp
#include <memory>

#include "../include/ImputeNoRef.h"
#include "../include/VCF.h"
#include "../include/VCFGeno.h"
#include "../include/SampleManager.h"
#include "../include/LargeFamily.h"
#include "../include/LargeSelfFamily.h"
#include "../include/materials.h"
#include "../include/SmallFamily.h"
#include "../include/option.h"

using namespace std;


//////////////////// ImputeNoRef ////////////////////

void ImputeNoRef::display_chromosome_info(const VCFSmall *orig_vcf) {
	cerr << "chr : " << orig_vcf->get_records().front()->chrom() << " ";
	if(orig_vcf->size() == 1) {
		cerr << "1 record." << endl;
	}
	else {
		cerr << orig_vcf->size() << " records." << endl;
	}
}

VCFGeno *ImputeNoRef::impute_vcf_chr(const VCFSmall *orig_vcf,
										SampleManager *sample_man,
										const Map& geno_map,
										const Option& option) {
	display_chromosome_info(orig_vcf);
	
	const auto	large_families = sample_man->get_large_families();
	auto	merged_vcf = LargeFamily::impute(orig_vcf, large_families,
															geno_map, option);
	if(merged_vcf != NULL)
		sample_man->add_imputed_samples(merged_vcf->get_samples());
	
	const auto	self_families =
						sample_man->extract_self_parent_non_imputed_families();
	auto	*imputed_vcf = LargeSelfFamily::impute(orig_vcf, merged_vcf,
											self_families, geno_map, option);
	if(imputed_vcf != NULL) {
		merged_vcf = imputed_vcf;
		sample_man->add_imputed_samples(merged_vcf->get_samples());
	}
	
	merged_vcf = SmallFamily::impute_small_family(orig_vcf, merged_vcf,
												geno_map, option, sample_man);
	
	sample_man->clear_imputed_samples();
	return merged_vcf;
}

void ImputeNoRef::impute(VCFHuge *vcf, const Materials *materials,
												const Option& option) {
	const vector<string>&	samples = vcf->get_samples();
	std::unique_ptr<const PedigreeTable> ped = nullptr;
	try {
		ped.reset(materials->get_ped()->limit_samples(samples));
	}
	catch(const ExceptionWithCode& e) {
		throw;
	}
	
	SampleManager	*sample_man = SampleManager::create(ped.get(), samples,
														vector<string>(),
														option.lower_progs,
														option.families);
	
	sample_man->display_info();
	
	// process chromosome by chromosome
	VCFHuge::ChromDivisor	divisor(vcf);
	bool	first_chromosome = true;
	for(int	chrom_index = 0; ; ++chrom_index) {
		// chrom_index is required only for because they can skip chromosomes
		const VCFSmall	*vcf_chrom = divisor.next();
		if(vcf_chrom == NULL)
			break;
		try {
			vcf_chrom->check_records();
		}
		catch(const std::exception& e) {
			delete vcf_chrom;
			delete sample_man;
			throw;
		}
		if(!option.is_efficient_chrom(chrom_index)) {
			delete vcf_chrom;
			continue;
		}
		
		const Map	*gmap = materials->get_chr_map(chrom_index);
		const VCFGeno	*vcf_imputed = impute_vcf_chr(vcf_chrom,
													sample_man, *gmap, option);
		if(first_chromosome) {
			ofstream	ofs(option.path_out);
			vcf_imputed->write(ofs, true);	// write header
		}
		else {
			ofstream	ofs(option.path_out, ios_base::app);
			vcf_imputed->write(ofs, false);
		}
		first_chromosome = false;
		delete vcf_imputed;
		delete vcf_chrom;
	}
	
	delete sample_man;
}

File: python/ImputeNoRef.py
from __future__ import annotations

# coding: utf-8
# ImputeNoRef.py
# リファレンス無しでimputeする

from itertools import *
import sys
from typing import Iterator, Optional

from VCF import *
from VCFGeno import VCFGeno
from pedigree import PedigreeTable, Family
import LargeFamily
import LargeSelfFamily
import SmallFamily
from materials import *
from SampleManager import SampleManager
from Map import *
from option import *
from OptionSmall import OptionSmall
from common import *


#################### ImputeNoRef ####################

def impute_vcf_chr(orig_vcf: VCFSmall, sample_man: SampleManager,
							geno_map: Map, option: Option) -> Optional[VCFGeno]:
	print('chr: %s %d records' % (orig_vcf.records[0].chrom(), len(orig_vcf)))
	sys.stdout.flush()
	merged_vcf = LargeFamily.impute(orig_vcf, sample_man.large_families,
															geno_map, option)
	if merged_vcf is not None:
		sample_man.set(merged_vcf.samples)
	
	self_families = sample_man.extract_self_parent_non_imputed_families()
	imputed_vcf = LargeSelfFamily.impute(self_families, orig_vcf, merged_vcf,
															geno_map, option)
	if imputed_vcf:
		merged_vcf = imputed_vcf
		sample_man.set(imputed_vcf.samples)
	
	op_small = OptionSmall(geno_map, option.num_threads, option.precision_ratio,
											option.imputes_isolated_samples,
											option.outputs_unimputed_samples)
	
	# 補完できる家系がなくなるまで繰り返す
	if merged_vcf:
		sample_man.set(merged_vcf.samples)
		merged_vcf = SmallFamily.impute(orig_vcf, merged_vcf,
												op_small, sample_man,
												option.imputes_isolated_samples)
	
	sample_man.clear()
	return merged_vcf

def impute(vcf: VCFHuge, materials: Materials, option: Option) -> None:
	samples = vcf.samples
	ped = materials.ped.limit_samples(samples)
	sample_man = SampleManager.create(ped, samples, [],
										option.lower_progs, option.families)
	sample_man.display_info(sys.stderr)
	
	iter = option.chroms_efficients()
	first = True
	for b, vcf_chr, gmap in zip(iter, vcf.divide_into_chromosomes(),
													materials.iter_chr_maps()):
		if not b:
			continue
		vcf_chr.check_records()
		vcf_imputed = impute_vcf_chr(vcf_chr, sample_man, gmap, option)
		if vcf_imputed is None:
			continue
		with open(option.path_out, 'w' if first else 'a') as out:
			vcf_imputed.write(out, with_header=first)
		first = False

__all__ = ['impute']
